### DISCUSSION TITLE
[Feature] Mixed Hardware Acceleration Capabiltity Expanded Microservices/Machine Learning Container Environment

### DISCUSSION THREAD

**Discussion by tbelway** [immich-app/immich#13648] [Feature Request]
### I have searched the existing feature requests to make sure this is not a duplicate request.

- [X] Yes

### The feature

Currently the video transcoding settings are configured in the administration GUI, where the Acceleration API can be chosen amongst a list including NVENC, Quick Sync, VAAPI, RKMPP and Disabled. However, it is possible one will have a mixed environment with, for example, an NVIDIA GPU passed through to one box, an intel CPU with iGPU on another box, and an AMD CPU on another.

It would be beneficial to allow for granular environment settings per container, so that you can have multiple microservices/ml nodes operating on these different environments and not have the transaction fail due to the GUI setting having NVENC, when it is trying to run on a box with Quick Sync or no possible hardware acceleration.

I would propose the option be set in GUI with a container level environment variable to override allowing for the primary instance having, for example, NVENC, and then each other instance can toggle between HWA being disables or enabled with a different HWA configuration.

:edit:

Proposed Requirements:
1) Either Partial Config Modification or Environment Variables
1a) If partial config modification, should take any undeclared values from database and not from default values
2) If Partial Config is detected, should not overwrite current values in DB
3) machine learning URL should accommodates multiple values, perhaps using an array instead of a string
3a) Need a way to to handle machine learning designation, either through redis or postgres?

### Platform

- [X] Server
- [ ] Web
- [ ] Mobile

**Comment by mertalev** [comment]
You can use config files to specify different transcoding settings in each container.

**Reply by tbelway** [reply]
Hey there @mertalev ,

Are you suggesting that, in a containerized environment, the config should be passed through to the container? The proposal I made was to use environment variables to pass through to the container, which would be the standard for container configuration. I don't see why this issue is closed given that the feature request has not had it's requirements met...

**Reply by mertalev** [reply]
We don't use environmental variables for settings in the admin UI because it gets confusing as to what setting is in active use, whether changing a setting in the UI will actually change what setting is in use, what to display in the UI to begin with, etc. Config files are the main way to deploy immich with a set configuration outside of the admin UI.

**Reply by tbelway** [reply]
I have a few concerns with what you have described.
1) We don't use environmental variables for settings in the admin UI because it gets confusing as to what setting is in active use
I could understand this if it was for the primary containers, but this would be easily remedied by having it explicitly stated in the admin UI that this is the default and it can be overwritten by environment variables or defined configs.
2) Config files are the main way to deploy immich with a set configuration outside of the admin UI
Immich is primarily deployed through containerization, and the standard is to use environment variables. It seems like breaking the standard and forcing passed through config files not only breaks the standard/philosophy of containerization, it makes k8s/k3s management much more difficult. An important part of scaling containerized environments includes using environment variables to facilitate the portability of these containers, and the use case I presented is an example as to why.

Currently, there is no documentation for running a multiple HWA, multiple container environment, and the recommendation you are providing is counter intuitive given the focus on containerization. Let's use another example as to why a managed config file is less flexible:
Let's say I want to utilize the GUI to manage my immich instance. I hope into the administration settings, and ensure that NVENC is the HWA API utilized. I then set up a second set of microserices/ml containers, and they rely on CPU. I therefore export my config and pass that through to my new instances. Uh oh, I now want to change the tone-mapping in the GUI from HABLE to MOBIUS.
Now, in order to ensure this is proper on both sets of containers I need to export the config file again, throw it on the new containerized hosts, and it'll apply. However, if I was using environment variables I could define the HWA API to disabled for the CPU based host, and when I make changes in the GUI all the other settigns will be reflected across all containers.

:edit:

This expands to virtually all settings in the GUI, too. Let's say I want to change the audio codec, or video codec, or the preset, or the CRF, or the target resolution... With your proposed flow, there is a constant requirement to both make the changes in the GUI as well as the managed configs for the other set of containers. Meanwhile, the environment variable is seamless, and all changes made in the config are propagated to all containers, maintaining the override in the environment variable.

**Reply by tbelway** [reply]
Also, @mertalev thank you for taking the time responding, you're super prompt and I appreciate it :D

**Reply by bo0tzz** [reply]
> Now, in order to ensure this is proper on both sets of containers I need to export the config file again, throw it on the new containerized hosts, and it'll apply. However, if I was using environment variables I could define the HWA API to disabled for the CPU based host, and when I make changes in the GUI all the other settigns will be reflected across all containers. 

The config file is partial, so you can have it be just overrides for the HWA settings and it should take everything else from the database.

**Reply by tbelway** [reply]
OK, so why is there resistance to the implementation of an environment variable which would allow for k8s or k3s deployments?

**Reply by bo0tzz** [reply]
Because we already have the config file, which works just fine with kubernetes as well.

**Reply by tbelway** [reply]
I have tested this, and it does not work as you have described.
Admin settings:
```
CLIP:
XLM-Roberta-Large-ViT-H-14__frozen_laion5b_s13b_b90k

Machine Learning URL:
http://${SANITIZED_FQDN}:3003
```

microservices1 (NVENC):
```
[Nest] 20  - 10/20/2024, 1:02:46 PM     LOG [Microservices:ServerService] Feature Flags: {
  "smartSearch": true,
  "facialRecognition": true,
  "duplicateDetection": true,
  "map": true,
  "reverseGeocoding": true,
  "importFaces": false,
  "sidecar": true,
  "search": true,
  "trash": true,
  "oauth": true,
  "oauthAutoLaunch": true,
  "passwordLogin": true,
  "configFile": false,
  "email": false
}
```

microservices2 (CPU):
```
[Nest] 2  - 10/22/2024, 4:08:28 PM     LOG [Microservices:EventRepository] Initialized websocket server
[Nest] 2  - 10/22/2024, 4:08:29 PM     LOG [Microservices:MapRepository] Initializing metadata repository
[Nest] 2  - 10/22/2024, 4:08:29 PM     LOG [Microservices:MetadataService] Initialized local reverse geocoder
[Nest] 2  - 10/22/2024, 4:08:30 PM     LOG [Microservices:SmartInfoService] Dimension size of model ViT-B-32__openai is 512, but database expects 1024.
[Nest] 2  - 10/22/2024, 4:08:30 PM     LOG [Microservices:SmartInfoService] Updating database CLIP dimension size to 512.
[Nest] 2  - 10/22/2024, 4:08:31 PM     LOG [Microservices:SmartInfoService] Successfully updated database CLIP dimension size from 1024 to 512.
[Nest] 2  - 10/22/2024, 4:08:31 PM     LOG [Microservices:SystemConfigService] LogLevel=log (set via system config)
[Nest] 2  - 10/22/2024, 4:08:31 PM     LOG [Microservices:MapRepository] Initializing metadata repository
[Nest] 2  - 10/22/2024, 4:08:31 PM     LOG [Microservices:MetadataService] Initialized local reverse geocoder
[Nest] 2  - 10/22/2024, 4:08:31 PM     LOG [Microservices:ServerService] Feature Flags: {
  "smartSearch": true,
  "facialRecognition": true,
  "duplicateDetection": true,
  "map": true,
  "reverseGeocoding": true,
  "importFaces": false,
  "sidecar": true,
  "search": true,
  "trash": true,
  "oauth": false,
  "oauthAutoLaunch": false,
  "passwordLogin": true,
  "configFile": true,
  "email": false
}
...
[Nest] 2  - 10/22/2024, 4:15:30 PM   ERROR [Microservices:JobService] Error: Machine learning request to "http://immich-machine-learning:3003" failed with Error: getaddrinfo ENOTFOUND immich-machine-learning
```

microservices2 config:
```
cat immich-config.json 
{
  "ffmpeg": {
    "accel": "disabled"
  }
}
```

Note the incorrect CLIP model, and the incorrect machine learning endpoint.

**Reply by mertalev** [reply]
I believe it would use the default settings for the fields not in the config file. It wouldn't use the database.

**Reply by tbelway** [reply]
> I believe it would use the default settings for the fields not in the config file. It wouldn't use the database.

That's what it does indeed look like based on my testing.
So... what would be the way forward here? Looks like regardless, work needs to be done for this feature request.

:edit:

The truly annoying part is it seems like it sets the entire config to default and pushes it to the database, so now I have some fixing to do...

**Reply by mertalev** [reply]
I'm not super against having environmental variables for these settings, but I think there needs to be more transparency in the settings UI around what is being used, where and why. That should happen regardless, as it's clearly not transparent enough as it is.

> The truly annoying part is it seems like it sets the entire config to default and pushes it to the database, so now I have some fixing to do...

Does it? It should only update the DB if you save in the admin settings.

**Reply by tbelway** [reply]
No changes were made in the admin settings, and yet the errors in the primary microservices container had to do with the model size which only differs from the default and the one I set. Which means it actually changed the value in the db...

In order to fix it I had to change it in admin settings to something different, and then back to my usual model.

Being a DBA and a unix systems administrator by career, I SHOULD have known better and created a temporary table first... but alas...

**Reply by mertalev** [reply]
Ah, I see. The CLIP model is special and gets checked at startup to make sure the DB is ready to accept embeddings from that model. In this case, the containers have different values for the model and compete to make the DB ready for their model. You'll need to re-run smart search. Sorry about that 😅

**Reply by tbelway** [reply]
Yup, I'm rerunning smart search now with the cuda image. It'll be a few days for this to complete... I have 1.5TB of photos. 

That being said, if this feature is implemented already then... what exactly is the correct way to do it? Clearly it's not as simple as outlined in this discussion...

I'd be happy to test as much as needed to g et it working, and I can even assist with the documentation.

:edit:

Well, holy crap, I guess you guys really optimized the CUDA ML. It's already at only 26k remaining! That means it'll probably be done in under an hour...

**Reply by tbelway** [reply]
So then, in order to launch multiple ML and microservices machines, what parameters need to defined?
Should we reopen this feature request?

**Reply by mertalev** [reply]
I reopened the FR.

Any non-default settings you want to apply to all jobs should be specified in each config file, including the CLIP model. Settings that should be different per instance, like transcoding, can have different values.

**Reply by tbelway** [reply]
OK, so the work around until the FR has traction is full configs then :) 

:edit:

Tested, at the very least the ML does not function in this way, the CPU instance is not getting any work at all. I assume it has to do with the ML API pointer not sharing a URL.
Microservices however is functioning as expected.

**Reply by tbelway** [reply]
Looks like today's release addresses this at least for hte most part :D

**Comment by tbelway** [comment]
Given the capabilities now exist, should close this discussion.
